### DISCUSSION TITLE
[Enhancement] Make easier to write view models

### DISCUSSION THREAD

**Discussion by danim-hub** [dotnet/maui#68] [General]
## Summary
Creating View Models requires following the INotifyPropertyChanged interface and related patterns. This distracts the developer from doing actual business logic to wiring events and writing getters/setters, losing the benefits of automatic properties.
Doing https://docs.microsoft.com/en-us/dotnet/api/xamarin.forms.bindableobject?view=xamarin-forms is too much work compared to an automatic property.

## API Changes
Add a new Attribute for the classes to be bound to UI, or perhaps extend the current BindableObject.

### Option1 using an Attribute:
```c#
using Microsoft.Maui.Stuff;
[Bindable] // new API, this make all properties 
public class CustomerViewModel : INotifyPropertyChanged {
    public string Id {get;set;}
    public string Name {get;set;}
}
```

### Option2 using an enhanced base class:
```c#
using Microsoft.Maui.Stuff;
public class CustomerViewModel : BindableObject {
    public string Id {get;set;}
    public string Name {get;set;}
    public CustomerViewModel {
        base.MakeBindable(this); // new API for enabling the property changed handling
    }
}
```

### Option3 using a wrapper class:
```c#
/// extension method
namespace Microsoft.Maui.Stuff {
public class BindingContextExtensions {
    public T ToBindingContext<T>(this T object) 
         where T:class
    {
        // TODO: subclass and wrap object T
    }
}
}

using Microsoft.Maui.Stuff;
public class CustomerViewModel {
    public string Id {get;set;}
    public string Name {get;set;}
    public CustomerViewModel {
        base.MakeBindable(this); // new API for enabling the property changed handling
    }
}

public class CustomerPage : Page {
    public CustomerPage {
        this.BindingContext = (new CustomerViewModel () ).ToBindingContext();
    }
}


```

## Intended Use Case
This will help us create less complicated ViewModels and reduce coding, helping developers adhere to the mvvm

**Comment by dhindrik** [comment]
I am using PropertyChanged.Fody for this. The only thing I need to do is to install the NuGet package. It will ad code by compile time.

**Comment by haavamoa** [comment]
I don't like magic and do not like to have to opt out property changed. PropertyChanged should be something you explicitly say that you want and not something you would have to turn off on my opinion. I don't think it's a good way of developing view models to always add those property changes. After all, you would not always want your properties to do that. If the view only updates the view model i.e


Leave this as is and let people choose to opt in autogeneration by 3rd parties.

**Comment by danim-hub** [comment]
You are right, you should probably have to set the attribute for each
property instead of each class.

Il giorno mer 20 mag 2020 alle 17:51 Håvard Moås <notifications@github.com>
ha scritto:

**Comment by m5x** [comment]
I like the attribute way (explicitly on properties, not the whole class). If it was at the same time possible to override generated `OnPropertyChanged` method with custom implementation it would be very elegant solution IMHO.

Example with a hypothetical notification manager being used to support collecting of notifications to dispatch them later etc.:

```csharp
public class CustomerViewModel : INotifyPropertyChanged
{
  [Bindable]
  public string Name { get; set; }
	
  [Bindable]
  public string Address { get; set; }
	
  // if this method exists then it gets called on every bindable property change instead of the generated one
  void OnPropertyChanged( string propertyName )
  {
    void NotifyPropertyChanged( string pn ) => PropertyChanged?( pn );

    if( _notificationManager != null )
      _notificationManager.Add( this, propertyName, NotifyPropertyChanged );
    else
      NotifyPropertyChanged( propertyName );
    }
}
```

**Comment by velocitysystems** [comment]
I think view model implementation should definitely be flexible, but would be in favour of deprecating `BindableObject`. View model implementation is highly opinionated and seems that Maui should really leave this as open as possible to the developer.

**Comment by ChaseFlorell** [comment]
We've never been forced to use `BindableObject` for our ViewModels, in fact you don't even have to use `INotifyPropertyChanged` if you're not doing any UI updates. I personally like the flexibility of the BindingContext being a simple `object` and allowing us to do whatever approach feels right.

The original options if implemented would cause ALL properties to notify, which isn't always what we want, we only want properties that interact with the View to update, for this there are already options out there.

Property Changed Fody, Reactive.Fody, BindableObject, Prism, MVVM Helpers, etc.

**Comment by yannduran** [comment]
I'd suggest being able to use `[Bindable]` on **both**:
-  on the _class_ to make **all** properties to provide property changed notifications
- on _properties_, so a developer can **choose** to not have all properties sending notifications

@dhindrik As far as I'm aware `PropertyChanged.Fody` produces IL for  ALL properties? I'm not using it, so I don't know for sure.

**Comment by mrwcjoughin** [comment]
I'm with @yannduran - [Bindable] attribute would be awesome! then I can use automatic properties and have them be bindable - what a joy that would be!

**Comment by SimonCropp** [comment]
@yannduran 

> As far as I'm aware PropertyChanged.Fody produces IL for ALL properties? I'm not using it, so I don't know for sure.

PropertyChanged processes properties of classes that implement `INotifyPropertyChanged` or have the marker attribute `[AddINotifyPropertyChangedInterface]`.

yes that may include some public properties that dont need notification, but firing a property event when nothing is using it is not a measurable performance impact in this context

**Comment by yannduran** [comment]
@SimonCropp 

Sorry, yes of course I meant that it produces IL for properties of classes that implement `INotifyPropertyChanged`.

Thanks for the confirmation.

**Comment by SimonCropp** [comment]
not sure if this interest anyone https://github.com/nicolo-ottaviani/Xamarin.BindableProperty.Fody

**Comment by haavamoa** [comment]
I use snippets for BindableProperty and PropertyChanged, problem solved 👍

**Comment by MikeKenyon** [comment]
At a **maximum**, I'd suggest having something like Bindable as an _option_.  Here's my concern.  In large projects there are a number of cases where bindability and responsiveness are broader concerns than represented in this sample:

- [ ]  One way, one-time and one-way-to-source bindings 
- [ ]  A property change conditionally creating an update (initialization mode for example)
- [ ]  A property update signaling a change in multiple properties (setting a combined name field updates GivenName and Surname)
- [ ]  A property update signaling a change to a different property (Updating a Type property shifts how a different property is calculated requiring that to be rebound)
- [ ]  A property update signaling a change from a method or event handler (parent object's children signals that the collection has changed, should result in a rebinding of the entire object).

**Comment by MhAllan** [comment]
I am against this, but if this goes anywhere the name should be [TriggersPropertyChange] which is also ugly :). 

Best is you do base class that wraps setting the property and triggers change event and then write snippet for that:

```csharp
public class ObservableObject : INotifyPropertyChanged, INotifyPropertyChanging
    {
        public event PropertyChangedEventHandler PropertyChanged;
        public event PropertyChangingEventHandler PropertyChanging;

        protected bool SetProperty<T>(ref T prop, T value, [CallerMemberName]string propertyName = "")
        {
            if (EqualityComparer<T>.Default.Equals(prop, value))
                return false;

            OnPropertyChanging(propertyName);
            prop = value;
            OnPropertyChanged(propertyName);
            return true;
        }

        protected bool SetProperty<T>(object obj, string prop, T value, [CallerMemberName] string propertyName = "")
        {
            var propInfo = obj.GetType().GetProperty(prop);
            var val = (T)propInfo.GetValue(obj);
            if (EqualityComparer<T>.Default.Equals(val, value))
                return false;

            OnPropertyChanging(propertyName);
            propInfo.SetValue(obj, value);
            OnPropertyChanged(propertyName);
            return true;
        }

        protected virtual void OnPropertyChanging([CallerMemberName]string propertyName = "")
        {
            PropertyChanging?.Invoke(this, new SC.PropertyChangingEventArgs(propertyName));
        }

        protected virtual void OnPropertyChanged([CallerMemberName]string propertyName = "")
        {
            PropertyChanged?.Invoke(this, new SC.PropertyChangedEventArgs(propertyName));
        }
    }
```

```xml
<CodeSnippet Format="1.0.0">
    <Header>
        <Title>INotifyPropertyChanged Snippet</Title>
        <Shortcut>propn</Shortcut>
      </Header>
    <Snippet>
      <Declarations>
        <Literal>
          <ID>type</ID>
          <Default>int</Default>
        </Literal>
        <Literal>
          <ID>p</ID>
          <Default>MyProperty</Default>
        </Literal>
      </Declarations>
      <Code Language="CSharp">
          <![CDATA[
          $type$ _$p$;
          public $type$ $p$
          {
            get { return _$p$; }
            set { SetProperty(ref _$p$, value); }
          }]]>
        </Code>
    </Snippet>
  </CodeSnippet>
```
